MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/BoundingBox.java

```diff
@@ -0,0 +1,124 @@
+
+/**
+ * A BoundingBox object represents the bounding box for a quadrilateral, i.e. the box enclosing the
+ * quadrilateral with the smallest measure.
+ */
+public class BoundingBox {
+
+    private static final String TAG = "BoundingBox";
+
+    /**
+     * The extreme coordinates of the bounding box.
+     */
+    public final float xMin;
+    public final float yMin;
+    public final float xMax;
+    public final float yMax;
+
+    /**
+     * A pseudo-measurement of how well the bounding box approximates the quadrilateral. This is
+     * computed by taking the maximum distance from a vertex of the quadrilateral to a neighboring
+     * vertex of the bounding box. Thus, the error lies in the range [0, max(W, H) / 2], where
+     * W is the width of the bounding box and H is the height of the bounding box.
+     */
+    private final float error;
+
+    /**
+     * Constructor for a bounding box using the bounding box coordinates.
+     * @param xMin The x-coordinate of the lower-left vertex of the bounding box.
+     * @param yMin The y-coordinate of the lower=left vertex of the bounding box.
+     * @param xMax The x-coordinate of the upper-right vertex of the bounding box.
+     * @param yMax The y-coordinate of the upper-right vertex of the bounding box.
+     */
+    public BoundingBox(float xMin, float yMin, float xMax, float yMax) {
+        this.xMin = xMin;
+        this.yMin = yMin;
+        this.xMax = xMax;
+        this.yMax = yMax;
+        error = 0.0f;
+    }
+
+    /**
+     * Constructor for a bounding box using a given quadrilateral.
+     * The arguments a, b, c, d are all homogeneous 2D coordinates (x, y, 1) of the vertices of the
+     * quadrilateral.
+     */
+    public BoundingBox(float[] a, float[] b, float[] c, float[] d) {
+        xMin = Math.min(Math.min(a[0], b[0]), Math.min(c[0], d[0]));
+        yMin = Math.min(Math.min(a[1], b[1]), Math.min(c[1], d[1]));
+        xMax = Math.max(Math.max(a[0], b[0]), Math.max(c[0], d[0]));
+        yMax = Math.max(Math.max(a[1], b[1]), Math.max(c[1], d[1]));
+
+        float errorA = Math.min(Math.abs(a[0] - xMin), Math.abs(a[0] - xMax));
+        if (errorA < EPS) {
+            errorA = Math.min(Math.abs(a[1] - yMin), Math.abs(a[1] - yMax));
+        }
+
+        float errorB = Math.min(Math.abs(b[0] - xMin), Math.abs(b[0] - xMax));
+        if (errorB < EPS) {
+            errorB = Math.min(Math.abs(b[1] - yMin), Math.abs(b[1] - yMax));
+        }
+
+        float errorC = Math.min(Math.abs(c[0] - xMin), Math.abs(c[0] - xMax));
+        if (errorC < EPS) {
+            errorC = Math.min(Math.abs(c[1] - yMin), Math.abs(c[1] - yMax));
+        }
+
+        float errorD = Math.min(Math.abs(d[0] - xMin), Math.abs(d[0] - xMax));
+        if (errorD < EPS) {
+            errorD = Math.min(Math.abs(d[1] - yMin), Math.abs(d[1] - yMax));
+        }
+
+        error = (float) Math.sqrt(2.0) *
+                Math.max(Math.max(errorA, errorB), Math.max(errorC, errorD));
```

Comment:
What will you use this error metrics for? It seems that it can still be very small (around `EPS`) for some quadrilateral that are not at all well approximated by their bounding box.

Reply:
So part of auto-crop was to re-center and re-scale the entire stabilized video so that the edges of the frames don't show in the viewport (without auto-crop, the video is stabilized but you can see the frame moving around in the viewport). I used BoundingBoxes to approximate the frames so that it would be easier to calculate how much I would need to translate and scale the whole video so that I could avoid seeing the edges of the frame as much as possible (and have the video fill up the whole view). But for a frame that is super rotated (like a diamond), the bounding box is a pretty bad approximation. So I calculated an error that (very loosely) approximated the longest distance from any vertex of the bounding box to an edge of the actual frame, and I use the error to increase the scale by some factor, just to hedge the possibility of bad approximations a little bit. It didn't work perfectly for all videos, but it did actually help in some super shaky videos.

It makes the auto-crop process a little more complicated though. Do you think it's worth keeping, or should I just stick with a simple approach?

Reply:
Up to you! I guess you can keep this heuristic if if works in practice, but I just wanted to mention that I definitely found cases where your metrics would return a very small value (indicating that the fit is good) even though it's really not. Once again, since we're out of time, feel free to not do anything here, but maybe add in the Javadoc that it's a approximate heuristic, not valid in all cases.

Reply:
Ok I think I see the cases that you are talking about (e.g. thin diagonal rectangle with all vertices close to two opposite vertices of the bounding box). I think I'll get rid of the error measure then. Maybe something like using the difference in area between the bounding box and the quadrilateral would have been better? But it seems like a good move to remove it for now since it wasn't adding too much benefit.

Reply:
Difference in area sounds like a good metrics to me :)